### PR TITLE
Augment Dockerfile for DockerHub

### PR DESCRIPTION
**Pull request by mabelzhang**
There were a few preferences I want to iterate on this PR about.

Currently this PR adds a new Dockerfile in a new directory. I didn't replace the original one, in case we are using it for development. @quarkytale how heavy do you depend on the rocker features? The biggest one I can see is the `--home` arg, which mounts your host directory. That can be done with pure rocker, but I didn't want to disrupt your workflow. Let me know if you think we can replace the rocker Dockerfile, and I can edit that instead of adding a new one.

Summary of changes:
- New `run.bash` without dependency on `rocker`
   Since the plan is to release the image on DockerHub, when people discover an image on DockerHub and `docker pull` it, they should be able to just `docker run` it, without having to install a third-party tool like `rocker`. We can’t expect people to look for a README, install rocker on their host, and then use that to run the Docker image. That adds a lot of overhead. While `rocker` is a handy tool for developers familiar with it, for the general public, raw `docker run` is a more familiar standard tool and more powerful. Everything that `rocker` does can be already done by `docker`.
- Consolidate `apt update` occurrences in Dockerfile, because each apt-get voids the previous cache and creates a bigger resulting image. For a smaller image, reduce `apt update`s, and each one should be followed by an `apt clean`
- Add non-root user in Dockerfile (used to be shortcut by rocker). Put source code in `/home/developer` as opposed to `/tmp`.

Notes on NVIDIA:
- run.bash
   There are two run scripts, one passes in NVIDIA, the other doesn’t. People without NVIDIA cards can fall back to the latter, but it will be slower (I thought it wasn’t too much slower, but I was only testing simple worlds). Both work for me right now. I have an NVIDIA card.
   I need someone else to test to make sure both work for them. Without NVIDIA, `.gz/rendering/ogre2.log` should show that it's using llvmpipe for software rendering. With NVIDIA, that file should show it's using NVIDIA.

- Base image
   For Gazebo to work in Dockerfiles, we usually default to basing on an [NVIDIA image](https://hub.docker.com/r/nvidia/opengl), such as `FROM nvidia/opengl:1.0-glvnd-devel-ubuntu20.04` (there isn’t an official one for 22.04, but I have a custom working version).
   Since Gazebo Garden, it’s no longer required to have an NVIDIA card for rendering in Docker. If there’s no NVIDIA card, Gazebo will use software rendering based on Mesa Llvmpipe. The caveat is that some people have found that they need to specify some environment variables for it to work. On my laptop, it just worked without specifying anything extra, and I didn’t notice any slowdown.
   For now, I’m leaving the base image the way it is (no NVIDIA), in the hopes that it is accessible to people without NVIDIA cards. If that doesn't work for others (e.g. cannot launch GUI even with environment variables, which we can add to the image), then I can change to basing on an NVIDIA image.

### CODE REVIEW

**Comment by quarkytale** [comment]
Yes definitely you can edit the original Dockerfile, it's not a very heavy dependence on rocker. I would just suggest to keep the bash scripts for users not familiar with docker won't have to deal with the raw docker tags. Removing NVIDIA seems fine too, any case not covered by this configuration would be revealed once we get a user base.

**Comment by mabelzhang** [comment]
It is now editing the existing Dockerfile. The rocker run.bash has been renamed run_rocker.bash.

I've removed all references to NVIDIA. Let's see if this works for you. If not, I can add NVIDIA.
If you have a computer without an NVIDIA card, that's especially good to test.

File: docker/buoy/Dockerfile
Comment (quarkytale) [inline docker/buoy/Dockerfile:67]: What's this file for? For running the simulation by default when going for the non-dev option?

Also, might need to update readme and install tutorial as well with the new instruction.

File: docker/run_rocker.bash
Comment (quarkytale) [inline docker/run_rocker.bash:1]: Can be deleted

File: docker/build.bash
Comment (quarkytale) [inline docker/build.bash:48]:
```suggestion
echo "./run.bash $image_name:latest"
```
Or update with the new usage

File: docker/run.bash
Comment (quarkytale) [inline docker/run.bash:62]: Remove debug comments

File: docker/run.bash
Comment (quarkytale) [inline docker/run.bash:38]: I didn't get this part here, what should I input this argument, `buoys_ws`?

File: docker/buoy/Dockerfile
Comment (mabelzhang) [reply docker/buoy/Dockerfile:67]: This is a shortcut for people to run the simulation. Someone pulling the Docker image from DockerHub will not have ready access to the README, and they'd have to dig through the files in the image. Having a shortcut enables them to quickly see what the Dockerfile does.

I didn't put it as an entrypoint, because we expect the users to be controls researchers, so they'll probably be developers writing plugins, and might not want the simulator to run every time they go into the container.

README - yeah will need to update.

File: docker/run.bash
Comment (mabelzhang) [reply docker/run.bash:38]: This variable is automatically assigned. The command to run should just be `./run.bash buoy`

File: docker/run_rocker.bash
Comment (mabelzhang) [reply docker/run_rocker.bash:1]: The shebang? This line is present in all the shell scripts. If it is run from a different shell, like sh, it tells sh to use bash. I would leave it in.

**Comment by mabelzhang** [comment]
7ce8715 is a major change

After thinking about it, I think the no-NVIDIA setup is too slow for regular use. It’s probably safer to offer an NVIDIA option, rather than waiting till things don’t work for people, to come back to add it.

I added an NVIDIA option `--no-nvidia` in `build.bash`, so the user can choose whether to build with an NVIDIA base image.

Same for `run.bash`.

While updating the README, I realized that having NVIDIA / no-NVIDIA options alongside the rocker instructions is too confusing, because to use rocker, which injects NVIDIA, they should build without NVIDIA. Then that creates 3 options, raw docker with NVIDIA, raw docker without NVIDIA, and rocker which should use the build option without NVIDIA. I think that’s too confusing.

With the current bash scripts, there isn’t really a reason to keep the rocker instructions around, because the home folder, NVIDIA, and non-root user were the main things rocker provided. These are already provided by the bash scripts.

Other changes

- `apt-key` had a deprecation warning. That is now replaced by the latest installation instructions on ROS and Gazebo websites.

- One improvement that I can make in a future PR is to use named layers, so that if not using NVIDIA, can base on the ros:humble-ros-base as before, and save image build time. Right now, all the lines need to be run, whether using NVIDIA or not, so the image takes longer to build than before.

File: docker/buoy/Dockerfile
Comment (mabelzhang) [reply docker/buoy/Dockerfile:117]: Updated in 7ce8715

File: docker/run_rocker.bash
Comment (quarkytale) [reply docker/run_rocker.bash:1]: No, I meant the entire file, or are we keeping rocker stuff in for now?

Comment (mabelzhang) [reply docker/run_rocker.bash]: Oh! Didn't see the title of that file. Yes should be removed.
Removed in 4ee2f1c

**Comment by quarkytale** [comment]
--no-nvidia works as expected, yet to test without the tag

**Review by quarkytale — APPROVED** [review]
Works for me with both the options, good to go after last touches.

File: README.md
Comment (mabelzhang) [reply README.md:155]: Oh I didn't realize the `join.bash` is a different one from the one I usually use. I've updated the file. It automatically finds the container ID. You should be able to join using `./join.bash buoy`. I've updated the README.
79b627f

File: docker/build.bash
Comment (mabelzhang) [reply docker/build.bash]: baadcf3

File: README.md
Comment (mabelzhang) [reply README.md]: Let me know if that works for you, before I merge

**Review by quarkytale — APPROVED** [review]

**Comment by quarkytale** [comment]
Merging this in, hoping there weren't any touches left

**Pull request merged**
